File: varint/varint.py
def encode(number: int) -> bytes:
    byte_array = []
    while number > 0:
        res = number & 0x7F
        number = number >> 7
        if number == 0:
            byte_array.append(res)
        else:
            byte_array.append(res + 0x80)
    return bytes(byte_array)


def decode(bytes_array: bytes) -> int:
    result = 0
    shift = 0
    for byte in bytes_array:
        byte = byte & 0x7F
        result += byte << shift
        shift = shift + 7
    return result


assert encode(150) == b"\x96\x01"
assert decode(b"\x96\x01") == 150
for n in range(1 << 30):
    assert decode(encode(n)) == n

File: varint/varint.rs
fn encode(mut number: u64) -> Vec<u8> {
    let mut bytes: Vec<u8> = Vec::new();
    while number > 0 {
        let res: u8 = (0x7F & number).try_into().unwrap();
        number = number >> 7;
        if number == 0 {
            bytes.push(res);
        } else {
            bytes.push(res + 0x80);
        }
    }
    return bytes;
}

fn decode(mut bytes: Vec<u8>) -> u64 {
    let mut result: u64 = 0;
    bytes.reverse();
    for &byte in bytes.iter() {
        result >>= 7;
        result |= byte as u64;
    }
    return result;
}
fn main() {
    assert_eq!(encode(150), vec![0x96, 0x1]);
    assert_eq!(decode(vec![0x96, 0x1]), 150);
    for number in 1..1 << 30 {
        assert_eq!(decode(encode(number)), number);
    }
}
